From 22b5e7ec70bdf9192476af7465d7eff2c34ab158 Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Tue, 22 Mar 2022 17:04:28 +0100
Subject: [PATCH 156/841] feat(st-fmc): add RIF support

The FMC2 revision 2 supports security and isolation compliant with
the Resource Isolation Framework (RIF). From RIF point of view,
the FMC2 is composed of several independent resources, listed below,
which can be assigned to different security and compartment domains:
 - 0: Common FMC_CFGR register.
 - 1: EBI controller for Chip Select 1.
 - 2: EBI controller for Chip Select 2.
 - 3: EBI controller for Chip Select 3.
 - 4: EBI controller for Chip Select 4.
 - 5: NAND controller.

FMC2 NAND controller has to be configured as secure to be able
to load binaries.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 3935649094e4c4ccf9e57d9a3e3e1c18f46d14dc feat(st-fmc): add RIF support]
Change-Id: I1948c4c36c2da458b0a6c61b0439bfbc12cdaee5
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/246020
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Christophe KERELLO <christophe.kerello@st.com>
---
 drivers/st/fmc/stm32_fmc2_nand.c | 97 +++++++++++++++++++++++++++++++-
 1 file changed, 96 insertions(+), 1 deletion(-)

diff --git a/drivers/st/fmc/stm32_fmc2_nand.c b/drivers/st/fmc/stm32_fmc2_nand.c
index 1f06eebef..1d5495de4 100644
--- a/drivers/st/fmc/stm32_fmc2_nand.c
+++ b/drivers/st/fmc/stm32_fmc2_nand.c
@@ -53,6 +53,11 @@
 #define FMC2_BCHDSR2			0x284U
 #define FMC2_BCHDSR3			0x288U
 #define FMC2_BCHDSR4			0x28CU
+#define FMC2_SECCFGR			0x300U
+#define FMC2_CIDCFGR0			0x30CU
+#define FMC2_CIDCFGR(x)			((x) * 0x8U + FMC2_CIDCFGR0)
+#define FMC2_SEMCR0			0x310U
+#define FMC2_SEMCR(x)			((x) * 0x8U + FMC2_SEMCR0)
 
 /* FMC2_BCR1 register */
 #define FMC2_BCR1_FMC2EN		BIT(31)
@@ -80,6 +85,7 @@
 #define FMC2_PCR_BCHECC			BIT(24)
 #define FMC2_PCR_WEN			BIT(25)
 /* FMC2_SR register */
+#define FMC2_SR_ISOST			GENMASK_32(1, 0)
 #define FMC2_SR_NWRF			BIT(6)
 /* FMC2_PMEM register*/
 #define FMC2_PMEM_MEMSET(x)		(((x) & GENMASK_32(7, 0)) << 0)
@@ -118,6 +124,16 @@
 #define FMC2_BCHDSR4_EBP7_MASK		GENMASK_32(12, 0)
 #define FMC2_BCHDSR4_EBP8_MASK		GENMASK_32(28, 16)
 #define FMC2_BCHDSR4_EBP8_SHIFT		16U
+/* FMC2_CIDCFGR register */
+#define FMC2_CIDCFGR_CFEN		BIT(0)
+#define FMC2_CIDCFGR_SEMEN		BIT(1)
+#define FMC2_CIDCFGR_SCID_MASK		GENMASK_32(6, 4)
+#define FMC2_CIDCFGR_SCID_SHIFT		4U
+#define FMC2_CIDCFGR_SEMWLC1		BIT(17)
+/* FMC2_SEMCR register */
+#define FMC2_SEMCR_SEM_MUTEX		BIT(0)
+#define FMC2_SEMCR_SEMCID_MASK		GENMASK_32(6, 4)
+#define FMC2_SEMCR_SEMCID_SHIFT		4U
 
 /* Timings */
 #define FMC2_THIZ			0x01U
@@ -126,6 +142,10 @@
 #define FMC2_PCR_TIMING_MASK		GENMASK_32(3, 0)
 #define FMC2_PMEM_PATT_TIMING_MASK	GENMASK_32(7, 0)
 
+#define FMC2_MAX_RESOURCES		6U
+#define FMC2_RESOURCE_CFGR		0U
+#define FMC2_RESOURCE_NAND		5U
+#define FMC2_CID1			1U
 #define FMC2_BBM_LEN			2U
 #define FMC2_MAX_ECC_BYTES		14U
 #define TIMEOUT_US_10_MS		10000U
@@ -667,6 +687,61 @@ static void stm32_fmc2_write_data(struct nand_device *nand,
 	}
 }
 
+#if STM32MP25
+static int stm32_fmc2_check_rif(unsigned int resource)
+{
+	uint32_t cidcfgr;
+	uint32_t semcr;
+	unsigned int cid;
+
+	if (resource >= FMC2_MAX_RESOURCES) {
+		return -EINVAL;
+	}
+
+	cidcfgr = mmio_read_32(fmc2_base() + FMC2_CIDCFGR(resource));
+	if ((cidcfgr & FMC2_CIDCFGR_CFEN) == 0U) {
+		/* CID filtering is turned off: access granted */
+		return 0;
+	}
+
+	if ((cidcfgr & FMC2_CIDCFGR_SEMEN) == 0U) {
+		/* Static CID mode */
+		cid = (cidcfgr & FMC2_CIDCFGR_SCID_MASK) >>
+		      FMC2_CIDCFGR_SCID_SHIFT;
+		if (cid != FMC2_CID1) {
+			ERROR("%s: static CID%u set for resource %u\n",
+			      __func__, cid, resource);
+			return -EACCES;
+		}
+
+		return 0;
+	}
+
+	/* Pass-list with semaphore mode */
+	if ((cidcfgr & FMC2_CIDCFGR_SEMWLC1) != 0U) {
+		ERROR("%s: CID1 is block-listed for resource %u\n",
+		      __func__, resource);
+		return -EACCES;
+	}
+
+	semcr = mmio_read_32(fmc2_base() + FMC2_SEMCR(resource));
+	if ((semcr & FMC2_SEMCR_SEM_MUTEX) != 0U) {
+		mmio_setbits_32(fmc2_base() + FMC2_SEMCR(resource),
+				FMC2_SEMCR_SEM_MUTEX);
+		semcr = mmio_read_32(fmc2_base() + FMC2_SEMCR(resource));
+	}
+
+	cid = (semcr & FMC2_SEMCR_SEMCID_MASK) >> FMC2_SEMCR_SEMCID_SHIFT;
+	if (cid != FMC2_CID1) {
+		ERROR("%s: resource %u is already used by CID%u\n",
+		      __func__, resource, cid);
+		return -EACCES;
+	}
+
+	return 0;
+}
+#endif
+
 static void stm32_fmc2_ctrl_init(void)
 {
 	uint32_t pcr = mmio_read_32(fmc2_base() + FMC2_PCR);
@@ -701,7 +776,16 @@ static void stm32_fmc2_ctrl_init(void)
 	mmio_setbits_32(fmc2_base() + FMC2_BCR1, FMC2_BCR1_FMC2EN);
 #endif
 #if STM32MP25
-	mmio_setbits_32(fmc2_base() + FMC2_CFGR, FMC2_CFGR_FMC2EN);
+	if ((mmio_read_32(fmc2_base() + FMC2_SR) & FMC2_SR_ISOST) != 0U) {
+		/* FMC2 is disabled */
+		if (stm32_fmc2_check_rif(FMC2_RESOURCE_CFGR) == 0) {
+			mmio_setbits_32(fmc2_base() + FMC2_CFGR,
+					FMC2_CFGR_FMC2EN);
+		} else {
+			/* FMC2 can not be enabled */
+			panic();
+		}
+	}
 #endif
 
 	mmio_write_32(fmc2_base() + FMC2_PCR, pcr);
@@ -937,6 +1021,17 @@ int stm32_fmc2_init(void)
 		}
 	}
 
+#if STM32MP25
+	/* Check if FMC2 NAND controller can be used */
+	if (stm32_fmc2_check_rif(FMC2_RESOURCE_NAND) == 0) {
+		/* Secure the FMC2 NAND controller */
+		mmio_setbits_32(fmc2_base() + FMC2_SECCFGR,
+				BIT(FMC2_RESOURCE_NAND));
+	} else {
+		panic();
+	}
+#endif
+
 	/* Setup default IP registers */
 	stm32_fmc2_ctrl_init();
 
-- 
2.47.2

